File: googlepay-base/src/main/java/com/adyen/checkout/googlepay/model/GooglePayPaymentMethodModel.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 17/7/2019.
 */

package com.adyen.checkout.googlepay.model;

import android.os.Parcel;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.model.JsonUtils;
import com.adyen.checkout.core.model.ModelObject;
import com.adyen.checkout.core.model.ModelUtils;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings("MemberName")
public class GooglePayPaymentMethodModel extends ModelObject {

    @NonNull
    public static final Creator<GooglePayPaymentMethodModel> CREATOR = new Creator<>(GooglePayPaymentMethodModel.class);

    private static final String TYPE = "type";
    private static final String PARAMETERS = "parameters";
    private static final String TOKENIZATION_SPECIFICATION = "tokenizationSpecification";

    @NonNull
    public static final Serializer<GooglePayPaymentMethodModel> SERIALIZER = new Serializer<GooglePayPaymentMethodModel>() {

        @NonNull
        @Override
        public JSONObject serialize(@NonNull GooglePayPaymentMethodModel modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                jsonObject.putOpt(TYPE, modelObject.getType());
                jsonObject.putOpt(PARAMETERS, ModelUtils.serializeOpt(modelObject.getParameters(), CardParameters.SERIALIZER));
                jsonObject.putOpt(TOKENIZATION_SPECIFICATION,
                        ModelUtils.serializeOpt(modelObject.getTokenizationSpecification(), PaymentMethodTokenizationSpecification.SERIALIZER));

            } catch (JSONException e) {
                throw new ModelSerializationException(GooglePayPaymentMethodModel.class, e);
            }
            return jsonObject;
        }

        @NonNull
        @Override
        public GooglePayPaymentMethodModel deserialize(@NonNull JSONObject jsonObject) {
            final GooglePayPaymentMethodModel googlePayPaymentMethodModel = new GooglePayPaymentMethodModel();
            googlePayPaymentMethodModel.setType(jsonObject.optString(TYPE, null));
            googlePayPaymentMethodModel.setParameters(ModelUtils.deserializeOpt(jsonObject.optJSONObject(PARAMETERS), CardParameters.SERIALIZER));
            googlePayPaymentMethodModel.setTokenizationSpecification(ModelUtils.deserializeOpt(jsonObject.optJSONObject(TOKENIZATION_SPECIFICATION),
                    PaymentMethodTokenizationSpecification.SERIALIZER));
            return googlePayPaymentMethodModel;
        }
    };

    private String type;
    private CardParameters parameters;
    private PaymentMethodTokenizationSpecification tokenizationSpecification;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public String getType() {
        return type;
    }

    public void setType(@Nullable String type) {
        this.type = type;
    }

    @Nullable
    public CardParameters getParameters() {
        return parameters;
    }

    public void setParameters(@Nullable CardParameters parameters) {
        this.parameters = parameters;
    }

    @Nullable
    public PaymentMethodTokenizationSpecification getTokenizationSpecification() {
        return tokenizationSpecification;
    }

    public void setTokenizationSpecification(@Nullable PaymentMethodTokenizationSpecification tokenizationSpecification) {
        this.tokenizationSpecification = tokenizationSpecification;
    }
}

File: googlepay/src/main/java/com/adyen/checkout/googlepay/model/GooglePayPaymentMethodModel.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 17/7/2019.
 */
package com.adyen.checkout.googlepay.model

import android.os.Parcel
import com.adyen.checkout.core.exception.ModelSerializationException
import com.adyen.checkout.core.model.JsonUtils.writeToParcel
import com.adyen.checkout.core.model.ModelObject
import com.adyen.checkout.core.model.ModelUtils.deserializeOpt
import com.adyen.checkout.core.model.ModelUtils.serializeOpt
import com.adyen.checkout.core.model.getStringOrNull
import org.json.JSONException
import org.json.JSONObject

data class GooglePayPaymentMethodModel(
    var type: String? = null,
    var parameters: CardParameters? = null,
    var tokenizationSpecification: PaymentMethodTokenizationSpecification? = null,
) : ModelObject() {
    override fun writeToParcel(dest: Parcel, flags: Int) {
        writeToParcel(dest, SERIALIZER.serialize(this))
    }

    companion object {
        private const val TYPE = "type"
        private const val PARAMETERS = "parameters"
        private const val TOKENIZATION_SPECIFICATION = "tokenizationSpecification"

        @JvmField
        val CREATOR = Creator(GooglePayPaymentMethodModel::class.java)

        @JvmField
        val SERIALIZER: Serializer<GooglePayPaymentMethodModel> = object : Serializer<GooglePayPaymentMethodModel> {
            override fun serialize(modelObject: GooglePayPaymentMethodModel): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(TYPE, modelObject.type)
                        putOpt(PARAMETERS, serializeOpt(modelObject.parameters, CardParameters.SERIALIZER))
                        putOpt(
                            TOKENIZATION_SPECIFICATION,
                            serializeOpt(
                                modelObject.tokenizationSpecification,
                                PaymentMethodTokenizationSpecification.SERIALIZER
                            )
                        )
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(GooglePayPaymentMethodModel::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): GooglePayPaymentMethodModel {
                val googlePayPaymentMethodModel = GooglePayPaymentMethodModel()
                googlePayPaymentMethodModel.type = jsonObject.getStringOrNull(TYPE)
                googlePayPaymentMethodModel.parameters = deserializeOpt(
                    jsonObject.optJSONObject(PARAMETERS),
                    CardParameters.SERIALIZER
                )
                googlePayPaymentMethodModel.tokenizationSpecification = deserializeOpt(
                    jsonObject.optJSONObject(TOKENIZATION_SPECIFICATION),
                    PaymentMethodTokenizationSpecification.SERIALIZER
                )
                return googlePayPaymentMethodModel
            }
        }
    }
}
